File: app/src/main/java/de/davis/passwordmanager/database/entities/details/password/PasswordDetails.java
package de.davis.passwordmanager.database.entities.details.password;

import java.io.Serial;
import java.util.Objects;

import de.davis.passwordmanager.database.ElementType;
import de.davis.passwordmanager.database.entities.details.ElementDetail;
import de.davis.passwordmanager.security.Cryptography;

public class PasswordDetails implements ElementDetail {

    @Serial
    private static final long serialVersionUID = 4938873580704485021L;

    private byte[] password;
    private String origin;
    private String username;
    private Strength strength;

    public PasswordDetails(String password, String origin, String username) {
        setPassword(password);
        this.origin = origin;
        this.username = username;
    }

    public String getOrigin() {
        return origin;
    }

    public void setOrigin(String origin) {
        this.origin = origin;
    }

    public String getUsername() {
        return username;
    }

    public void setUsername(String username) {
        this.username = username;
    }

    public Strength getStrength() {
        return strength;
    }

    public void setPassword(String password){
        this.password = Cryptography.encryptAES(password.getBytes());
        this.strength = Strength.estimateStrength(password);
    }

    public byte[] getPasswordData() {
        return password;
    }

    public String getPassword(){
        return new String(Cryptography.decryptAES(getPasswordData()));
    }

    @Override
    public ElementType getElementType() {
        return ElementType.PASSWORD;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        PasswordDetails that = (PasswordDetails) o;
        return getPassword().equals(that.getPassword()) && Objects.equals(origin, that.origin) && Objects.equals(username, that.username);
    }

    @Override
    public int hashCode() {
        return Objects.hash(origin, username, getPassword());
    }
}

File: app/src/main/java/de/davis/passwordmanager/database/entities/details/password/PasswordDetails.kt
package de.davis.passwordmanager.database.entities.details.password

import com.google.gson.annotations.SerializedName
import de.davis.passwordmanager.database.ElementType
import de.davis.passwordmanager.database.entities.details.ElementDetail
import de.davis.passwordmanager.security.Cryptography
import java.io.Serial

class PasswordDetails(
    password: String,
    var origin: String,
    var username: String,
) : ElementDetail {

    var strength: Strength = EstimationHandler.estimate(password)
        private set

    @SerializedName("password")
    private var passwordEncrypted: ByteArray = Cryptography.encryptAES(password.toByteArray())

    fun setPassword(password: String) {
        strength = EstimationHandler.estimate(password)
        passwordEncrypted = Cryptography.encryptAES(password.toByteArray())
    }

    val password get() = String(Cryptography.decryptAES(passwordEncrypted))


    override fun getElementType(): ElementType {
        return ElementType.PASSWORD
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as PasswordDetails

        if (origin != other.origin) return false
        if (username != other.username) return false
        if (!passwordEncrypted.contentEquals(other.passwordEncrypted)) return false

        return true
    }

    override fun hashCode(): Int {
        var result = origin.hashCode()
        result = 31 * result + username.hashCode()
        result = 31 * result + passwordEncrypted.contentHashCode()
        return result
    }


    companion object {
        @Serial
        private val serialVersionUID = 4938873580704485021L
    }
}